### PR TITLE
Feat implement state for new changelogs

### PR DESCRIPTION
**Pull request by ntorionbearstudio**
Please make sure the below checklist is followed for Pull Requests.
  - [ ] Checks are green
  - [ ] Tests are added where necessary
  - [ ] Documentation is added/updated where necessary
  - [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/jhipster-core/blob/master/CONTRIBUTING.md) are followed

https://github.com/jhipster/generator-jhipster/issues/11398

### CODE REVIEW

File: lib/exporters/jhipster_entity_exporter.js [hunk 1]
```diff
@@ -119,6 +119,10 @@ function writeEntities(subFolder) {
   for (let i = 0, entityNames = Object.keys(configuration.entities); i < entityNames.length; i++) {
     const filePath = path.join(subFolder, toFilePath(entityNames[i]));
     const entity = updateEntityToGenerateWithExistingOne(filePath, configuration.entities[entityNames[i]]);
+    if (FileUtils.doesFileExist(filePath)) {
+      const destPath = path.join(subFolder, toPreviousStateFilePath(entityNames[i]));
```
Comment (MathieuAA) [inline lib/exporters/jhipster_entity_exporter.js:123]: Please use the full word: `destinationPath`

File: lib/exporters/jhipster_entity_exporter.js [hunk 2]
```diff
@@ -20,7 +20,7 @@
 const fs = require('fs');
 const path = require('path');
 const ApplicationTypes = require('../core/jhipster/application_types');
-const { toFilePath, readJSONFile } = require('../readers/json_file_reader');
+const { toFilePath, toPreviousStateFilePath, readJSONFile } = require('../readers/json_file_reader');
```
Comment (MathieuAA) [inline lib/exporters/jhipster_entity_exporter.js:23]: I don't get what the function does just by looking at the name.
What does it do?

File: yarn.lock [hunk 3]
```diff
@@ -0,0 +1,5393 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (MathieuAA) [inline yarn.lock:1]: Please remove this file. We use NPM

File: lib/readers/json_file_reader.js [hunk 4]
```diff
@@ -60,3 +61,15 @@ function toFilePath(entityName) {
   }
   return path.join('.jhipster', `${upperFirst(entityName)}.json`);
 }
+
+/**
+ * From an entity's name, gives the expected previous state file path.
+ * @param entityName the entity's name.
+ * @returns {string} the file's path.
+ */
+function toPreviousStateFilePath(entityName) {
```
Comment (MathieuAA) [inline lib/readers/json_file_reader.js:70]: `getFilePathForPreviousStateOfEntity` would be a better name

Don't hesitate to be verbose!

File: lib/readers/json_file_reader.js [hunk 5]
```diff
@@ -60,3 +61,15 @@ function toFilePath(entityName) {
   }
   return path.join('.jhipster', `${upperFirst(entityName)}.json`);
 }
+
+/**
+ * From an entity's name, gives the expected previous state file path.
+ * @param entityName the entity's name.
```
Comment (MathieuAA) [inline lib/readers/json_file_reader.js:67]: {String} missing

File: lib/readers/json_file_reader.js [hunk 6]
```diff
@@ -60,3 +61,15 @@ function toFilePath(entityName) {
...
+ * @returns {string} the file's path.
```
Comment (MathieuAA) [inline lib/readers/json_file_reader.js:68]: {**S**tring}

File: lib/exporters/jhipster_entity_exporter.js (hunk 2 above)
Comment (MathieuAA) [reply lib/exporters/jhipster_entity_exporter.js:23]: I've suggested another name, perhaps it's more suggestive

File: lib/exporters/jhipster_entity_exporter.js [hunk 7]
```diff
@@ -119,6 +119,10 @@ function writeEntities(subFolder) {
   for (let i = 0, entityNames = Object.keys(configuration.entities); i < entityNames.length; i++) {
     const filePath = path.join(subFolder, toFilePath(entityNames[i]));
     const entity = updateEntityToGenerateWithExistingOne(filePath, configuration.entities[entityNames[i]]);
+    if (FileUtils.doesFileExist(filePath)) {
+      const destPath = path.join(subFolder, toPreviousStateFilePath(entityNames[i]));
+      fs.copyFileSync(filePath, destPath);
+    }
```
Comment (MathieuAA) [inline lib/exporters/jhipster_entity_exporter.js:125]: What does this block do?

**Comment by MathieuAA** [comment]
Tests are missing BTW

**Comment by mshima** [comment]
@MathieuAA what about returning the last entity state instead of creating a new file?
We can pass the structure to the generators and the generator/blueprint save there instead, if required.

I don’t like the idea of creating -previous file by default

**Comment by MathieuAA** [comment]
@mshima I like your idea. @ntorionbearstudio is @mshima idea compatible with your feature?

**Comment by MathieuAA** [comment]
@ntorionbearstudio Hello there! This repo's source code is being moved to the generator, and this PR's additions too.
The next steps for me are:
  - either finish the PR here and copy/paste the changes to another PR in the generator once https://github.com/jhipster/generator-jhipster/pull/11694 is merged
  - or wait for the PR to be merged (https://github.com/jhipster/generator-jhipster/pull/11694) and continue your work after that

I'll help you finish this PR if need be

**Comment by ntorionbearstudio** [comment]
> @mshima I like your idea. @ntorionbearstudio is @mshima idea compatible with your feature?

Yes ! I'm working on it this week :)

**Comment by pascalgrimaud** [comment]
Thanks @ntorionbearstudio but be carefull. This repo will be archived, because jhipster-core has been merged into generator-jhipster

**Comment by ntorionbearstudio** [comment]
Ok I'm working on updated generator-jhipster repo

**Comment by ntorionbearstudio** [comment]
I moved what I needed on the generator-jhipster repo : https://github.com/jhipster/generator-jhipster/pull/11419

Now the last entities states are passed to `entity` generator in param and no more `-previous` files are created as @mshima suggested.

So I'm closing this PR and continue to work on https://github.com/jhipster/generator-jhipster/pull/11419

**Pull request closed without merging**